homeassistant/components/homepluscontrol/api.py

```diff
@@ -0,0 +1,348 @@
+        hass (HomeAssistant): HomeAssistant core object.
+        config_entry (ConfigEntry): ConfigEntry object that configures this API.
+        implementation (AbstractOAuth2Implementation): OAuth2 implementation that handles AA and
+                                                       token refresh.
+        logger (Logger): Logger of the object.
+        switches (dict): Dictionary of HomePlusControl switches indexed by their unique ID.
+        switches_to_remove (dict): Dictionaty of the HomePlusControl switches that are to be
+                                   removed from HomeAssistant, indexed by their unique ID.
+    """
+
+    def __init__(
+        self,
+        hass: core.HomeAssistant,
+        config_entry: config_entries.ConfigEntry,
+        implementation: config_entry_oauth2_flow.AbstractOAuth2Implementation,
+    ):
+        """Initialize the HomePlusControlAsyncApi object.
+
+        Initialize the authenticated API for the Legrand Home+ Control component.
+
+        Args:.
+            hass (HomeAssistant): HomeAssistant core object.
+            config_entry (ConfigEntry): ConfigEntry object that configures this API.
+            implementation (AbstractOAuth2Implementation): OAuth2 implementation that handles AA
+                                                           and token refresh.
+        """
+        self.hass = hass
+        self.config_entry = config_entry
+        self.implementation = implementation
+        self._domain = DOMAIN
+        self._plants = {}
+        self._switches = {}
+        self._switches_to_remove = {}
+
+        self._last_check = {
+            CONF_PLANT_UPDATE_INTERVAL: time.monotonic(),
+            CONF_PLANT_TOPOLOGY_UPDATE_INTERVAL: -1,
+            CONF_MODULE_STATUS_UPDATE_INTERVAL: -1,
+        }
+
+        self._refresh_intervals = {}
+        self.update_refresh_intervals()
+
+        # Create the API authenticated client - external library
+        super().__init__(
+            client_id=self.config_entry.data[CONF_CLIENT_ID],
+            client_secret=self.config_entry.data[CONF_CLIENT_SECRET],
+            subscription_key=self.config_entry.data[CONF_SUBSCRIPTION_KEY],
+            redirect_uri=self.config_entry.data[CONF_REDIRECT_URI],
+            token=self.config_entry.data["token"],
+            oauth_client=aiohttp_client.async_get_clientsession(self.hass),
+        )
+
+    @property
+    def logger(self) -> logging.Logger:
+        """Logger of authentication API."""
+        return logging.getLogger(__name__)
+
+    @property
+    def switches(self):
+        """Return dictionary of switch entities of this platform."""
+        return self._switches
+
+    @switches.setter
+    def switches(self, switches):
+        """Set the internal switch attribute."""
+        self._switches = switches
+
+    @property
+    def switches_to_remove(self):
+        """Return dictionary of switch entities of this platform that should be removed from HA."""
```

Comment:
This api should not know anything about the Home Assistant `Entity` instances. This is the device api and should only handle the devices. The entity platforms will then consume this api when there is new information, triggered by the coordinator update.

Reply:
I realized that the wording was confusing here. The api is not really handling Home Assistant entities, but only the integrations "switch modules". I made the change of calling them "modules" instead of "entities".